bytecomp/translmod.ml

```diff
@@ -247,23 +247,17 @@ let init_shape modl =
 
 (* Reorder bindings to honor dependencies.  *)
 
-type binding_status = Undefined | Inprogress | Defined
-
-let extract_unsafe_cycle id fv cycle_start =
-  let num_bindings = Array.length id in
-  let rec explore = function
-    | [] -> assert false
-    | (l,i) :: stack ->
-        if List.mem i l then
-          List.rev_map (Array.get id) (i::l)
-        else begin
-          let children =
-            List.filter
-              ( fun (_,j) -> Ident.Set.mem id.(j) fv.(i) )
-            @@ List.init num_bindings (fun j -> i::l, j ) in
-          explore (children @ stack) end
-  in
-  explore [[], cycle_start]
+type binding_status =
+  | Undefined
+  | Inprogress of int option (** parent node *)
+  | Defined
+
+let extract_unsafe_cycle id status cycle_start =
```

Comment:
You could have a comment on the fact that `cycle_start` will occur both at start and the end of the result list. This is a desirable property in the end-user printing, but it is not immediate (to me) that it is the natural specification of `extract_unsafe_cycle`, and someone reading this code without looking at the test could think the duplication is a mistake.

(Or a version of this function without the repetition would be fine (not clearly better); I would then expect the currently-repeated element to be present only at the end of the list, to make it easy to add the duplicate back afterwards with just a cons.)

Reply:
I added a comment on this duplication, since indeed this not a completely natural representation of a cycle.